cross-app-access-blog (#1603)
oktadeveloper/okta-blog: _source/_posts/2025-09-01-cross-app-access.md

---
layout: blog_post
title: "Build Secure Agent-to-App Connections with Cross App Access (XAA)"
author: sohail-pathan
by: advocate

> communities: [devops,security,mobile,.net,java,javascript,go,php,python,ruby]

Reviewer: ```suggestion
communities: [javascript]
```

description: "Learn how to securely connect agents to applications using Cross App Access (XAA) in this comprehensive guide."
tags: [xaa, cross-app-access, enterprise-ai, agentic-ai, mcp, oauth, sso]
tweets:
- ""
- ""
- ""
- ""
image: blog/cross-app-access/social.jpg
type: conversion
---

> Secure access between independent applications that communicate with each other is a recognized gap in [OAuth 2.0](https://developer.okta.com/docs/concepts/oauth-openid/). Now, with the advent of AI agents that communicate across systems, the need to solve the gap is even greater.

Reviewer: We need a link to describe the recognized gap. We might understand what the gap is, but no one else does. If we can't find a link to explain, then this sentence must be reworked. Linking to OAuth overview doesn't help.

[Cross App Access (XAA)](https://help.okta.com/oie/en-us/content/topics/apps/apps-cross-app-access.htm) is a new protocol that lets ISVs enable secure agent-to-app and app-to-app access. Instead of scattered integrations and repeated logins, enterprise IT admins gain centralized control: they can decide what connects, enforce security policies, and see exactly what's being accessed. This unlocks seamless, scalable integrations across apps — whether it's just two like Google Calendar and Zoom, or hundreds across the enterprise.

{% youtube 3VLzeT1EGrg %}

In this post, we'll go hands-on with Cross App Access. Using **Todo0** (Resource App) and **Agent0** (Requesting App) as our sample applications, and **Okta as the enterprise Identity Provider (IdP)**, we'll show you how to set up trust, exchange tokens, and enable secure API calls between apps. By the end, you'll not only understand how the protocol works but also have a working example you can adapt to your own integrations.

If you'd rather watch a video of the setup and how XAA works, check this one out. 
{% youtube vi5JpbGRATE %}

> ## Prerequisites

Reviewer: We don't make prereq a header as mentioned previously. please follow existing patterns.

Reviewer: Once again, use headers to reinforce SEO keywords. If you must have a header for pre-reqs, then add something about to contribute to SEO.

Author: Added

2. **AWS Credentials:** You'll need an **AWS Access Key ID** and **AWS Secret Access Key**.  
   * The IAM user or role associated with these credentials must have access to **Amazon Bedrock,** specifically the **Claude 3.7 Sonnet model,** enabled.  
   * If you don't know how to obtain the credentials, [follow this guide](https://github.com/oktadev/okta-cross-app-access-mcp/blob/main/guide/aws-bedrock.md).

3. **Developer Tools:** These tools are essential for cloning, editing, building, and running your demo applications:  
   * **[Git](https://git-scm.com/downloads)** – to clone and manage the repository.  
   * **[VS Code](https://code.visualstudio.com/Download)** – for reading and modifying the sample source code.  
   * **[Dev Containers Extension (VS Code)](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers)** – recommended, as it automatically configures dependencies and environments when you open the project.  
   * **[Docker](https://www.docker.com/products/docker-desktop/)** – required by the Dev Container to build and run the sample applications in isolated environments

**Table of Contents**{: .hide }
* Table of Contents
{:toc}

> ## Create applications in Okta Admin Console

Reviewer: ```suggestion
## Create applications in the Okta Admin Console
```

---
layout: blog_post
title: "Build Secure Agent-to-App Connections with Cross App Access (XAA)"
author: sohail-pathan
by: advocate
communities: [javascript]
description: "Learn how to securely connect agents to applications using Cross App Access (XAA) in this comprehensive guide."
tags: [xaa, cross-app-access, enterprise-ai, agentic-ai, mcp, oauth, sso]
tweets:
- ""
- ""
- ""
- ""
image: blog/cross-app-access/social.jpg
type: conversion
---

> Secure access between independent applications that communicate with each other is a recognized gap in [OAuth 2.0](https://developer.okta.com/docs/concepts/oauth-openid/). Now, with the advent of AI agents that communicate across systems, the need to solve the gap is even greater.

Reviewer: ```suggestion
Secure access with enterprise IT oversight between independent applications that communicate with each other is a recognized gap in [OAuth 2.0](https://developer.okta.com/docs/concepts/oauth-openid/). Enterprises can't effectively regulate cross-app communication, as OAuth 2.0 consent screens rely on users granting access to their individual accounts. Now, with the advent of AI agents that communicate across systems, the need to solve the gap is even greater.
```

Checking with Aaron

> [Cross App Access (XAA)](https://help.okta.com/oie/en-us/content/topics/apps/apps-cross-app-access.htm) is a new protocol that lets ISVs enable secure agent-to-app and app-to-app access. Instead of scattered integrations and repeated logins, enterprise IT admins gain centralized control: they can decide what connects, enforce security policies, and see exactly what's being accessed. This unlocks seamless, scalable integrations across apps — whether it's just two like Google Calendar and Zoom, or hundreds across the enterprise.

Reviewer: ```suggestion
Cross App Access (XAA) is a new protocol that lets integrators enable secure agent-to-app and app-to-app access. Instead of scattered integrations and repeated logins, enterprise IT admins gain centralized control: they can decide what connects, enforce security policies, and see exactly what's being accessed. This unlocks seamless, scalable integrations across apps — whether it's just two like Google Calendar and Zoom, or hundreds across the enterprise. Read more about Cross App access in this post:

{% excerpt /blog/2025/06/23/enterprise-ai %}
```
